compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt

```diff
@@ -472,19 +474,47 @@ abstract class FirDataFlowAnalyzer<FLOW : Flow>(
     ) {
         val leftIsNullable = leftOperand.coneType.isMarkedNullable
         val rightIsNullable = rightOperand.coneType.isMarkedNullable
+
         // left == right && right not null -> left != null
+        // [processEqNull] adds both implications: operator call could be true or false. We definitely need the matched case only.
+        fun operationStatementTester(operationStatement: OperationStatement, block: () -> Unit) {
+            // Only if operation statement is == True, i.e., left == right
+            if ((operation.isEq() && operationStatement.operation == Operation.EqTrue) ||
+                (!operation.isEq() && operationStatement.operation == Operation.EqFalse)
+            ) {
+                block()
+            }
+        }
         when {
             leftIsNullable && rightIsNullable -> return
-            leftIsNullable -> processEqNull(node, leftOperand, operation.invert())
-            rightIsNullable -> processEqNull(node, rightOperand, operation.invert())
+            leftIsNullable -> processEqNull(node, leftOperand, operation.invert(), ::operationStatementTester)
+            rightIsNullable -> processEqNull(node, rightOperand, operation.invert(), ::operationStatementTester)
         }
 
         if (operation == FirOperation.IDENTITY || operation == FirOperation.NOT_IDENTITY) {
             processIdentity(node, leftOperand, rightOperand, operation)
         }
     }
 
-    private fun processEqNull(node: EqualityOperatorCallNode, operand: FirExpression, operation: FirOperation) {
+    /*
+     * Process x == null in general: add implications for both cases.
+     * E.g., say d1 is an eq operator call node, d2 represents `x`, the variable inside the operator call. So, d1: x == null
+     * This util adds: "d1 == True -> d2 == null" and "d1 == False -> d2 != null"
+     * so that both branches after the operator call can infer the type of x.
+     *
+     * However, users can specify what conditions are of interest.
+     * E.g., say left == right _and_ right != null, then we can conclude left != null.
+     * In this example, say d1 is an eq operator call (left == right), and d2 is left.
+     * Unlike general cases, we want to add: "d1 == True -> d2 != null", and nothing more because the counter part,
+     * "d1 == False -> d2 == null" doesn't hold. That is, left != right and right != null don't mean left == null. It just means, left is
+     * something different from right, including null. By filtering "d1 == True" condition only, all the remaining logic can be shared.
+     */
+    private fun processEqNull(
+        node: EqualityOperatorCallNode,
+        operand: FirExpression,
+        operation: FirOperation,
+        operationStatementTester: (OperationStatement, () -> Unit) -> Unit = { _, block -> block() }
```

Comment:
This lambda looks strange. It will be cleaner if this lambda will have `(OperationStatement) -> Boolean` type.

Reply:
Check [rr/demiurg906/jsjeon/fir-null-smartcast](https://github.com/JetBrains/kotlin/tree/rr/demiurg906/jsjeon/fir-null-smartcast) branch, where I added this change

Reply:
Makes sense. Thanks for the branch with an update.